### PR TITLE
speed up ci

### PR DESCRIPTION
**Pull request by shamilovtim**
## Reason for PR
CI runs especially on the `test-with-browsers` job were taking between 15 to 25 minutes. Some common optimizations were unimplemented and parallelism was not being used for tests despite GH for open source being free.

## Implementation
- bump action versions
- add node cache 
- add playwright cache 
- replace docker with native `node` (see: https://github.com/TBD54566975/dwn-server/pull/94)
- parallelize browser tests
- only do the minimum amount of work necessary for the browser workflow (e.g. building cjs isnt necessary there)

## Results
- CI runs are now consistently between **230% to 380% faster**
- Duration variance of ~10 minutes has almost completely been eliminated. Runs now finish in a fairly consistent time frame. 
- CI duration went from **15min to 25min** (frequently 20min+) to **6min to 8min**

### CODE REVIEW

File: .github/workflows/alpha-npm.yml [hunk 1]
```diff
@@ -25,16 +25,14 @@ jobs:
 
     steps:
       - name: Checkout source
-        uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab # v3.5.2
+        uses: actions/checkout@v4
 
       - name: Set up Node.js
-        uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # v3.6.0
+        uses: actions/setup-node@v4
```
Comment (frankhinek) [inline .github/workflows/alpha-npm.yml:31]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: .github/workflows/docs-publish.yml [hunk 2]
```diff
@@ -25,13 +25,14 @@ jobs:
 
     steps:
       - name: Checkout source
-        uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab # v3.5.2
+        uses: actions/checkout@v4
```
Comment (frankhinek) [inline .github/workflows/docs-publish.yml:28]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: .github/workflows/docs-publish.yml [hunk 3]
```diff
@@ -71,7 +72,7 @@ jobs:
 
     steps:
       - name: Checkout
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
```
Comment (frankhinek) [inline .github/workflows/docs-publish.yml:75]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: .github/workflows/release-npm.yml [hunk 4]
```diff
@@ -37,16 +37,14 @@ jobs:
 
     steps:
       - name: Checkout source
-        uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab # v3.5.2
+        uses: actions/checkout@v4
```
Comment (frankhinek) [inline .github/workflows/release-npm.yml:40]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: .github/workflows/release-npm.yml [hunk 5]
```diff
@@ -37,16 +37,14 @@ jobs:
...
 
       - name: Set up Node.js
-        uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # v3.6.0
+        uses: actions/setup-node@v4
```
Comment (frankhinek) [inline .github/workflows/release-npm.yml:43]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: .github/workflows/scorecard.yml [hunk 6]
```diff
@@ -32,7 +32,7 @@ jobs:
 
     steps:
       - name: "Checkout code"
-        uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8 # v3.1.0
+        uses: actions/checkout@v4
```
Comment (frankhinek) [inline .github/workflows/scorecard.yml:35]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: .github/workflows/scorecard.yml [hunk 7]
```diff
@@ -59,7 +59,7 @@ jobs:
       # Upload the results as artifacts (optional). Commenting out will disable uploads of run results in SARIF
       # format to the repository Actions tab.
       - name: "Upload artifact"
-        uses: actions/upload-artifact@3cea5372237819ed00197afe530f5a7ea3e805c8 # v3.1.0
+        uses: actions/upload-artifact@v3
```
Comment (frankhinek) [inline .github/workflows/scorecard.yml:62]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: .github/workflows/tests-ci.yml [hunk 8]
```diff
@@ -16,15 +16,13 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout source
-        uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab # v3.5.2
+        uses: actions/checkout@v4
```
Comment (frankhinek) [inline .github/workflows/tests-ci.yml:19]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: .github/workflows/tests-ci.yml [hunk 9]
```diff
@@ -16,15 +16,13 @@ jobs:
...
 
       - name: Set up Node.js
-        uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # v3.6.0
+        uses: actions/setup-node@v4
```
Comment (frankhinek) [inline .github/workflows/tests-ci.yml:22]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: .github/workflows/tests-ci.yml [hunk 10]
```diff
@@ -50,16 +48,14 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout source
-        uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab # v3.5.2
+        uses: actions/checkout@v4
```
Comment (frankhinek) [inline .github/workflows/tests-ci.yml:51]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: .github/workflows/tests-ci.yml [hunk 11]
```diff
@@ -50,16 +48,14 @@ jobs:
...
 
       - name: Set up Node.js
-        uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # v3.6.0
+        uses: actions/setup-node@v4
```
Comment (frankhinek) [inline .github/workflows/tests-ci.yml:54]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: .github/workflows/tests-ci.yml [hunk 12]
```diff
@@ -70,65 +66,96 @@ jobs:
...
+      matrix:
+        include:
+          - group: "A"
+            packages: "--workspace packages/agent --workspace packages/dids "
+          - group: "B"
+            packages: "--workspace packages/common --workspace packages/crypto --workspace packages/credentials --workspace packages/user-agent --workspace packages/proxy-agent --workspace packages/api --workspace packages/identity-agent"
     steps:
       - name: Checkout source
-        uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab # v3.5.2
+        uses: actions/checkout@v4
```
Comment (frankhinek) [inline .github/workflows/tests-ci.yml:101]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: .github/workflows/tests-ci.yml [hunk 13]
```diff
@@ -70,65 +66,96 @@ jobs:
...
 
       - name: Set up Node.js
-        uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # v3.6.0
+        uses: actions/setup-node@v4
```
Comment (frankhinek) [inline .github/workflows/tests-ci.yml:104]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: .github/workflows/tests-ci.yml [hunk 14]
```diff
@@ -70,65 +66,96 @@ jobs:
...
-        run: npm run test:browser --ws
+      - name: Terminate dwn-server
+        run: kill $DWN_SERVER_BACKGROUND_PROCESS || true
 
   web5-spec:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout source
-        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        uses: actions/checkout@v4
```
Comment (frankhinek) [inline .github/workflows/tests-ci.yml:160]: [Security scanning](https://github.com/TBD54566975/web5-js/security/code-scanning) checks for pinned dependencies using hash so pin to a v4 using the commit hash of the GitHub action.

> For GitHub workflows used in building and releasing your project, pin dependencies by hash.

File: packages/agent/web-test-runner.config.cjs [hunk 15]
```diff
@@ -24,7 +24,7 @@ module.exports = {
   concurrentBrowsers : 2,
   testFramework      : {
     config: {
-      timeout: '15000',
+      timeout: '30000',
```
Comment (frankhinek) [inline packages/agent/web-test-runner.config.cjs:27]: Are there specific tests that are taking longer than 15 seconds to run?  Curious if there is just 1 or 2 outliers.

**Review by frankhinek — CHANGES_REQUESTED** [review]
Nice work @shamilovtim !  Will be great to speed up the test runs for future PRs.

File: packages/agent/web-test-runner.config.cjs (hunk 15 above)
Comment (shamilovtim) [reply packages/agent/web-test-runner.config.cjs:27]: its the usual suspects in the `agent` test suite

**Comment by shamilovtim** [comment]
resolved @frankhinek

**Review by frankhinek — APPROVED** [review]

**Pull request merged**